Sources/Rosalind/Rosalind.swift

```diff
@@ -1 +1,88 @@
+import Command
+import CryptoKit
+import FileSystem
+import Foundation
+import Path
 
+enum RosalindError: LocalizedError {
+    case notFound(AbsolutePath)
+
+    var errorDescription: String? {
+        switch self {
+        case let .notFound(path):
+            return "File not found at path \(path.pathString)"
+        }
+    }
+}
+
+public protocol Rosalindable {
+    func analyze(path: AbsolutePath) async throws -> Report
+}
+
+public struct Rosalind: Rosalindable {
+    let fileSystem: FileSysteming
+    let commandRunner: CommandRunning
+
+    public init() {
+        self.init(fileSystem: FileSystem(), commandRunner: CommandRunner())
+    }
+
+    init(
+        fileSystem: FileSysteming,
+        commandRunner: CommandRunning
+    ) {
+        self.fileSystem = fileSystem
+        self.commandRunner = commandRunner
+    }
+
+    public func analyze(path: AbsolutePath) async throws -> Report {
+        if !(try await fileSystem.exists(path)) { throw RosalindError.notFound(path) }
+        return try await traverse(path: path, normalizingPath: path)
+    }
+
+    public func traverse(path: AbsolutePath, normalizingPath: AbsolutePath) async throws -> Report {
+        let children: [Report] = try await (try await fileSystem.glob(directory: path, include: ["*"]).collect()).sorted()
+            .asyncMap { try await traverse(path: $0, normalizingPath: normalizingPath) }
+        let size = try await size(path: path, children: children)
+        let shasum = try await shasum(path: path, children: children)
+        if path.extension == "app" {
+            return .app(path: path.relative(to: normalizingPath).pathString, size: size, shasum: shasum, children: children)
+        } else {
+            return .unknown(path: path.relative(to: normalizingPath).pathString, size: size, shasum: shasum, children: children)
+        }
+    }
+
+    private func shasum(path: AbsolutePath, children: [Report]) async throws -> String {
+        if try await fileSystem.exists(path, isDirectory: true) { // Directory
+            let sortedHashes = children.map(\.shasum).sorted()
+            let combinedString = sortedHashes.joined()
+            guard let data = combinedString.data(using: .utf8) else {
+                return ""
```

Comment:
should we throw here instead?

Reply:
I think we can safely assume the conversion to UTF-8-encoded data is going to work. If we are wrong in the assumption, we can add some error-handling measures.